lsコマンド

k-yano-gs opened the pull request.

k-yano-gs committed:
lsコマンドを作成しました。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,23 @@
+end
+
+def print_files(row, col, parts)
+  (0..col - 1).each do |i|
```
`each` を使うと `-1` する必要が出てしまいますね。　`times` を使うともっとスッキリ書けるので、その方向で修正を検討してみてください 🙏

yuuu replied:
`parts` は事前に2次元配列化されているので、それを素直に `each` した方が `row` や `col` といったインデックスを意識しなくて良くなりますね。 `print_files` に渡すべき値も `parts` のみでよくなると思うのですが、その方向で修正できそうでしょうか？

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,23 @@
+def print_files(row, col, parts)
+  (0..col - 1).each do |i|
+    (0..row - 1).each do |j|
+      print (parts[j][i]).to_s.ljust(20) if !parts[j].nil?
```
`20` という数値は仕様を表す数字であり、仕様は変わる可能性があるので、変えやすいように定数としておきましょう。

yuuu replied:
ここは好みによるので修正するかどうかはお任せしますが、 `if` の代わりに `unless` を使うのも手です。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,23 @@
+
+def main
+  files = Dir.glob('*')
+  row = 3
```
プログラムの処理上は `3` という数字が途中まで行数を表すのですが、仕様としての `3` は列数なので変数名は `column` の方が読みやすいですね。ここも定数化しておきましょう。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,23 @@
+def main
+  files = Dir.glob('*')
+  row = 3
+  col = (files.size / row.to_f).ceil
```
この書き方でもOKですが、Ruby 3.2以降では [Integer#ceildiv](https://docs.ruby-lang.org/ja/latest/method/Integer/i/ceildiv.html) というメソッドがあります。これを使うと更にスッキリ書くことができるので、知っておいてください。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,23 @@
+  files = Dir.glob('*')
+  row = 3
+  col = (files.size / row.to_f).ceil
+  parts = files.each_slice(col).map do |part|
```
ここは以下のように書けますね。

```ruby
parts = files.each_slice(col).to_a
```

yuuu replied:
`part` という単語は「部品」という意味であり、 `file` よりも抽象的な名前になっていますね。`each_slice` したことで `files` に特徴が加わった変数という認識なので、変数名はより具体的な名前になることを期待します。

k-yano-gs committed:
commit2
colをcolumnにしました。
行と列の変数を入れ替えました。
Integer#ceildivメソッドを使うようにしました。
変数partの名前を変更しました。
print_filesdメソッの引数が１つでいいように繰り返し文を変更しました。
文字寄せの数値を変数にしました。
if文をunless文にしました。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,22 @@
+end
+
+def print_files(parts_files)
+arrange = 20
```
この書き方だと定数ではなく変数となっているので、定数化をお願いします 🙏 
https://docs.ruby-lang.org/ja/latest/doc/spec=2fvariables.html#const

定数はメソッドの外側(ファイルの先頭付近)で宣言すると良いですね。

yuuu replied:
定数を別途定義いただいたので、こちらの変数は削除をお願いします 🙏

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,22 @@
+
+def print_files(parts_files)
+arrange = 20
+  parts_files.each_with_index do |row,i|
```
ブロック変数 `row` が使われていないですが、こちらを活用すればコードが更にスッキリしますし、使用するメソッドも `each_with_index` ではなく `each` で済むはずです。

`column` についても同様に修正をお願いします 🙏

yuuu replied:
前回の段階では `transpose` を使わないために `each_with_index` を使っていた、ということだったんですね 💦 
`transpose` は終わっているものだと思い込んで上記コメントをしていました。

結果的には `transpose` を使うことで行列入れ替えができるということに気づいていただき、コードもスッキリしたので良かったとも思っています ✨

k-yano-gs committed:
commit3
文字寄せの数値を定数にしました。
parts_filseをtranseposeして縦に並べられるようにしました。
each_with_index文をeach文にしました。

k-yano-gs committed:
commit4
不要な文を削除しました。

k-yano-gs committed:
commit5
transposeの前に配列の要素の数を統一するようにしました
